birdvoxpaint/util.py

```diff
@@ -0,0 +1,225 @@
+
+    return y_blocks, sr
+
+
+# def multifile_stream(filenames, block_length):
+#     '''Chain multiple file audio block streams into a single stream. It uses samples
+#     from the next file to complete a previously incomplete block
+#     '''
+#     raise NotImplemented("I haven't really played around or tested this. I just drafted it.")
+#     remaining = None
+#     for filename in filenames:
+#         offset = 0 # TODO: uhhh I guess ? - idk
+#         # TODO: Load up to block_length -
+#         if remaining is not None:
+#             offset = block_length - len(remaining)
+#             y_complete, _ = librosa.load(filename, duration=offset)
+#             remaining = np.concatenate([remaining, y_complete])
+#
+#             if len(remaining) == block_length: # there were enough samples in the new file
+#                 yield remaining
+#                 remaining = None # clear
+#             else: # super short file ??? - better to be robust to all types of weather.
+#                 continue
+#
+#         y_blocks = librosa.stream(filename,
+#             offset=offset,
+#             block_length=block_n_frames,
+#             frame_length=frame_length,
+#             hop_length=hop_length)
+#
+#         for y in y_blocks:
+#             librosa.util.valid_audio(y, mono=True)
+#             if sr != orig_sr:
+#                 librosa.resample(y, orig_sr, sr)
+#             yield y
+
+
+
+def spec_frame(S, segment_length):
+    '''Break a spectrogram into equal segments.
+
+    Returns:
+        Spectrograms with the window applied
+        shape: (freq, frame * time) => (frame, freq, time)
+
+    '''
+
+    # only feed in full length frames into the pipeline
+    truncated_length = (S.shape[1] // segment_length) * segment_length
+    S = S[:,:truncated_length] # removes incomplete frames
+    S = S.T.reshape(-1, segment_length, len(S)) # shape: frame, segment, freq
+    S = S.transpose((0, 2, 1)) # shape: frame, freq, segment
+    return S
+
+
+
+def apply_indices(S, indices, segment_length, n_jobs=1):
+    '''Takes an iterable yielding blocks of spectrograms, breaks them into blocks of
+    size: segment_length. The number of blocks per yield should equal n_jobs.
+
+    Arguments:
+        S (iterable(np.ndarray[freq, time])): the iterable of spectrograms
+        indices (callable/list of callables): the indices to calculate. If a
+            single callable, it should return all indices as the final axis.
+        segment_length (int): the length in time (# frames) to be passed to
+            indices at a time.
+
+    '''
+    # reshape to blocks of spectrograms (# blocks == n_jobs)
+    S_blocks = (spec_frame(spec, segment_length) for spec in S) # yields: block, freq, time
+
+    # allow user to pass a custom indices function - or as a list of functions
+    if callable(indices):
+        calc_indices = indices
+    else:
+        calc_indices = lambda x: np.stack([f(x) for f in indices], axis=-1)
```

Comment:
i think you can remove the brackets `[]` here and make this list comprehension a generator. it will allocate less memory.
see this video by Raymond Hettinger if you're interested!
https://www.youtube.com/watch?v=OSGv2VnC0go

Reply:
I was getting warnings about using generators feeding into numpy functions 

```
In [6]: a = np.stack(([i]*10 for i in range(10)))                                                                                         
/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3249: FutureWarning: arrays to stack must be passed as a "sequence" type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
  if (await self.run_code(code, result,  async_=asy)):
```